### ISSUE TITLE
API Server is frequently re-deployed 

### ISSUE THREAD

**Issue by alzabo** [tigera/operator#1776]
The tigera-operator appears to be restarting the calico-apiserver pods every 10 minutes.

## Expected Behavior
Calico API Server does not restart unless a logical change is made.

## Current Behavior
calico-apiserver pods restart approximately every 10 minutes, as shown by the following output from `kubectl get pods -n calico-apiserver -w`

```
$ kubectl get pods -n calico-apiserver -w
NAME                                READY   STATUS    RESTARTS   AGE
calico-apiserver-6f4fbf7b54-d2dd6   1/1     Running   0          4m48s
calico-apiserver-6f4fbf7b54-mwmps   1/1     Running   0          4m48s
calico-apiserver-6f4fbf7b54-v2k2g   1/1     Running   0          4m48s
calico-apiserver-6f4fbf7b54-mwmps   1/1     Terminating   0          10m
calico-apiserver-6f4fbf7b54-d2dd6   1/1     Terminating   0          10m
calico-apiserver-6f4fbf7b54-v2k2g   1/1     Terminating   0          10m
calico-apiserver-6f4fbf7b54-hwsvd   0/1     Pending       0          1s
calico-apiserver-6f4fbf7b54-hwsvd   0/1     Pending       0          1s
calico-apiserver-6f4fbf7b54-cjw9h   0/1     Pending       0          0s
calico-apiserver-6f4fbf7b54-nlgzj   0/1     Pending       0          0s
calico-apiserver-6f4fbf7b54-cjw9h   0/1     Pending       0          0s
calico-apiserver-6f4fbf7b54-nlgzj   0/1     Pending       0          0s
calico-apiserver-6f4fbf7b54-hwsvd   0/1     ContainerCreating   0          1s
calico-apiserver-6f4fbf7b54-cjw9h   0/1     ContainerCreating   0          0s
calico-apiserver-6f4fbf7b54-nlgzj   0/1     ContainerCreating   0          0s
calico-apiserver-6f4fbf7b54-v2k2g   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-d2dd6   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-mwmps   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-d2dd6   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-d2dd6   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-cjw9h   0/1     Running             0          1s
calico-apiserver-6f4fbf7b54-mwmps   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-mwmps   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-nlgzj   0/1     Running             0          1s
calico-apiserver-6f4fbf7b54-hwsvd   0/1     Running             0          3s
calico-apiserver-6f4fbf7b54-v2k2g   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-v2k2g   0/1     Terminating         0          10m
calico-apiserver-6f4fbf7b54-cjw9h   1/1     Running             0          9s
calico-apiserver-6f4fbf7b54-nlgzj   1/1     Running             0          12s
calico-apiserver-6f4fbf7b54-hwsvd   1/1     Running             0          14s
```

When this happens, logs like the following are printed by the tigera-operator pod - The following were all printed in short order at the time when the apiserver deployment was restarted:
```
tigera-operator-648bcdc9cf-67cmw tigera-operator {"level":"info","ts":1644436641.1278439,"logger":"controller_apiserver","msg":"Reconciling APIServer","Request.Namespace":"","Request.Name":"default"}
tigera-operator-648bcdc9cf-67cmw tigera-operator {"level":"info","ts":1644436641.8467112,"logger":"controller_apiserver","msg":"Reconciling APIServer","Request.Namespace":"","Request.Name":"default"}
tigera-operator-648bcdc9cf-67cmw tigera-operator {"level":"info","ts":1644436644.8428574,"logger":"status_manager","msg":"update to tigera status conflicted, retrying","reason":"Operation cannot be fulfilled on tigerastatuses.operator.tigera.io \"apiserver\": the object has been modified; please apply your changes to the latest version and try again"}
tigera-operator-648bcdc9cf-67cmw tigera-operator {"level":"info","ts":1644436649.841111,"logger":"status_manager","msg":"update to tigera status conflicted, retrying","reason":"Operation cannot be fulfilled on tigerastatuses.operator.tigera.io \"apiserver\": the object has been modified; please apply your changes to the latest version and try again"}
tigera-operator-648bcdc9cf-67cmw tigera-operator {"level":"info","ts":1644436659.8435774,"logger":"status_manager","msg":"update to tigera status conflicted, retrying","reason":"Operation cannot be fulfilled on tigerastatuses.operator.tigera.io \"apiserver\": the object has been modified; please apply your changes to the latest version and try again"}
```

An event from the calico-apiserver namespace which appears related is:
```
11m         Warning   FailedToUpdateEndpoint   endpoints/calico-api                     Failed to update endpoint calico-apiserver/calico-api: Operation cannot be fulfilled on endpoints "calico-api": the object has been modified; please apply your changes to the latest version and try again
```

## Steps to Reproduce (for bugs)
Unfortunately, I'm not sure what a minimal number of steps to reproduce would be. I've provided some environment details below.

1.
2.
3.
4.

## Your Environment

* Operating System and version: Amazon Linux 2 (official EKS AMI v1.19.15-eks-9c63c4)
* Amazon EKS 1.19
* Some other operators in the cluster such as Flux and Kyverno are managing Calico NetworkPolicies and Calico HostEndpoints using the v3 CRDs, and so are hitting the API Server.
* Calico installed with tigera-operator helm chart, tigera-operator-v3.22.0
* Values are 
```
helm get values calico -n tigera-operator
USER-SUPPLIED VALUES:
apiServer:
  enabled: true
calicoctl:
  image: docker.io/calico/ctl
  tag: v3.22.0
certs:
  node:
    cert: null
    commonName: null
    key: null
  typha:
    caBundle: null
    cert: null
    commonName: null
    key: null
imagePullSecrets: {}
installation:
  componentResources:
  - componentName: Node
    resourceRequirements:
      limits:
        cpu: 60m
        memory: 200Mi
      requests:
        cpu: 10m
        memory: 200Mi
  - componentName: KubeControllers
    resourceRequirements:
      limits:
        cpu: 100m
        memory: 128Mi
      requests:
        cpu: 10m
        memory: 64Mi
  - componentName: Typha
    resourceRequirements:
      limits:
        cpu: 100m
        memory: 500Mi
      requests:
        cpu: 10m
        memory: 400Mi
  controlPlaneNodeSelector:
    node-type.services.myco.com: mgmt
  controlPlaneReplicas: 3
  controlPlaneTolerations:
  - effect: NoSchedule
    key: node.services.myco.com
    operator: Equal
    value: mgmt
  enabled: true
  kubernetesProvider: ""
  typhaAffinity:
    nodeAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
        nodeSelectorTerms:
        - matchExpressions:
          - key: node-type.services.myco.com
            operator: In
            values:
            - mgmt
tigeraOperator:
  image: tigera/operator
  registry: quay.io
  version: v1.25.0
```

**Comment by tmjd** [comment]
Sorry for not responding earlier.
From the multiple restarts my initial feeling is that the operator might need to sort some configuration that goes into the apiserver deployment that it deploys. We've seen an issue like this before where one Reconcile is done for the API server and renders the deployment in one order then the next Reconcile/render changes an order which causes the deployment to be updated and causes the deployment to restart/re-deploy the pods.

It would be interesting to see a `kubectl get deployment -n calico-apiserver -l k8s-app=calico-apiserver -o yaml` at multiple times to try to see if there is something in the deployment that is being reordered.

**Comment by caseydavenport** [comment]
Believe you can pass `-o yaml` to the watch command to see precisely what changes are being made

**Comment by wiswesser** [comment]
I just detected the same issue on our kubernetes cluster, and I believe I have found the reason for this seemingly strange behaviour.
TL;DR:
Flux and the Tigera-Operator keep changing something irrelevant back and forth (In our case a "/" was removed or added to a Registryname). This causes the Tigera-Operator to restart the apiserver every 10 minutes. To fix this, find out what the difference is and correct this in your Gitrepository.


At first I didn't even notice the many restarts of the apiserver and only noticed an increase in load every 10 minutes. This lead me to look at our flux configuration, because it reconciles the cluster every 10 Minutes (The default setting). Only while doing this I noticed the restarting calico-apiserver and believing to have found the actual culprit I started googling if this problem had occured before and found this thread. The fact that @alzabo also uses flux, tipped me off, and some testing confirmed that the restarts happen every time flux reconciles the cluster. It turned out that I had made a mistake when configuring our Installation CRD and was missing a slash in our custom registry url. This in itself was no problem, because the tigera-operator fixed this error for me. But now the cluster state and git repository were out of sync, causing flux to change the Installation CRD back every time it reconciled the cluster, which was of course fixed again by the tigera-operator and caused the restart. I fixed this by using the "flux diff" command to figure out, what flux is going to change and then correct the relevant value in our git repository.

I am not sure if anything needs to be changed in the tigera-operator code. Hopefully anyone else running into this (Probably very specific and rare) problem, finds this thread and the solution.
I'd argue that the operator should throw an error, instead of silently fixing the problem, so the user knows whats going on. But I can see the benefits of having the automatic fix.

What still confuses me, is why a change in the Installation CRD (Not the ApiServer CRD) causes the apiserver nodes to restart, but not all other running calico pods. Maybe someone else can explain why this is the case.

**Comment by caseydavenport** [comment]
Great sleuthing!

> What still confuses me, is why a change in the Installation CR (Not the ApiServer CR) causes the apiserver nodes to restart, but not all other running calico pods. Maybe someone else can explain why this is the case.

Yes, this is a good question. I would expect that this particular change would not cause any pods to restart, because the operator should always be adding that "/" and thus the rendered k8s resources should be identical even though the CR is changing. So, I think there's something more to learn here, but thanks for reporting back with your findings!

**Comment by tmjd** [comment]
>What still confuses me, is why a change in the Installation CR (Not the ApiServer CR) causes the apiserver nodes to restart, but not all other running calico pods. Maybe someone else can explain why this is the case.

I think the reason for this behavior is because the apiserver controller does not update/"fix" the Installation CR. It simply reads it and uses any fields it needs.

The controller for the installation/calico-system (called core if you look at the code) reads the Installation CR and "fixes" the registryname adding the `/` if it is missing. That same controller also is responsible for calico-node, typha, and calico-kube-controllers so those pods are never updated with the missing trailing `/`.

In other words the sequence is:
1. Installation CR gets updated removing a trailing slash from RegistryName
2. APIserver controller sees the change and starts applying the change, updating the apiserver deployment
3. Installation/calico-system controller sees the Installation CR and fixes the RegistryName, adding the trailing `/` back.
4. APIserver controller sees the Installation CR update with the trailing `/` and applies the change, updating the apiserver deployment back to what it was before step 2.

@alzabo Could this be the reason you're seeing this behavior too? If I'm reading the watch output right it looks like you see the restart every 10 minutes. Are you setting RegistryName also in your Installation CR managed by Flux?

**Comment by caseydavenport** [comment]
@tmjd so we just need the API server controller to also ensure the "/" is present before it calls render? Or should we remove the logic from the installation controller to write the "/" back to the API?

**Comment by mtheeren-asml** [comment]
Thanks a lot @wiswesser, @tmjd I was also struggling a while to see that it was indeed the trailing `/` that was causing the problems with redeployment here
